design/ceph-nfs-ganesha.md

```diff
@@ -0,0 +1,166 @@
+# Ceph NFS-Ganesha CRD
+
+[NFS-Ganesha] is a user space NFS server that is well integrated with [CephFS]
+and [RGW] backends. It can export Ceph's filesystem namespaces and Object
+gateway namespaces over NFSv4 protocol.Rook already orchestrates Ceph filesytem
+and Object store (or RGW) on Kubernetes. It can be extended to orchestrate
+NFS-Ganesha server daemons as highly available and scalable NFS gateway pods
+to the Ceph filesytem and Object Store. This will allow NFS client applications
+to use the Ceph filesytem and object store setup by rook.
+
+## Design of Ceph NFS-Ganesha CRD
+
+The NFS-Ganesha server settings will be exposed to rook as a Custom Resource
+Definition (CRD). Creating the nfs-ganesha CRD will launch NFS Ganesha server
+pods that will be custom configured for CephFS and RGW backends (e.g., Ganesha
+caching turned off). The servers' exports and client recovery data will be
+stored in Ceph RADOS pools to allow the Ganesha server pods to scale and be
+highly available. And the servers will have stable IP addresses.
+
+### Prerequisities
+
+- A running rook Ceph filesystem or object store, whose namespaces will be
+  exported by the Ganesha servers.
+  e.g.,
+  ```
+  kubectl create -f filesystem.yaml
+  ```
+
+- An existing RADOS pool (e.g., CephFS's data pool) or a pool created from
+  the [Rook toolbox] that will  store the Ganesha exports and NFS client
+  recovery data.
+
+- A RADOS object in the pool that will store the Ganesha exports, and an
+  initial export stored in the object that the server will start off exporting.
+  ```
+  # cat << EOF > ganesha-export.conf
+  EXPORT
+  {
+        Export_ID=100;
+        Protocols = 4;
+        Transports = TCP;
+        Path = /;
+        Pseudo = /cephfs/;
+        Access_Type = RW;
+        Attr_Expiration_Time = 0;
+        Delegations = R;
+        Squash = No_root_squash;
+        FSAL {
+                Name = CEPH;
+        }
+  }
+  EOF
+  # rados -p myfs-data0 put ganesha-export ganesha-export.conf
+  ```
+
+### Ceph NFS-Ganesha CRD
+
+The NFS-Ganesha CRD will specify the following:
+
+- RADOS pool where the Ganesha exports and NFS client recovery data will be
+  stored.
+
+- RADOS object in the pool that will have the Ganesha exports
+
+- The number of active Ganesha servers
+
+- The placement of the Ganesha servers
+
+- The resource limits (memory, CPU) of the Ganesha pods
+
+Below is an example NFS-Ganesha CRD, `nfs-ganesha.yaml`
+
+```yaml
+apiVersion: ceph.rook.io/v1alpha1
+kind: NFSGanesha
+metadata:
+  name: my-nfs
+  namespace: rook-ceph
+spec:
+  export:
+    # RADOS pool where Ganesha exports and NFS client recovery data is stored.
+    pool: myfs-data0
+    # RADOS object in the pool that has the Ganesha exports
+    object: ganesha-export
+  server:
+    # the number of active ganesha servers
+    active: 1
```

Comment:
StatefulSet uses "replicas:" for this. Should we follow suit?

Also, I assume we'll be able to scale up/down the cluster by changing the "active" value? I wonder if we need to allow for more granular ways to resize the cluster?

For instance, suppose we have a 5 node cluster and want to scale down to a 4 node cluster. With the StatefulSet controller, you can only eliminate ganesha-5 when scaling down the cluster size. Suppose ganesha-3 is relatively idle and ganesha-5 is busy. We'd have to migrate all of the clients onto ganesha-3 which might be more disruptive. Can we come up with a way that we could eliminate a particular server in the middle of the range from the set?

Maybe we need to consider a set of bitmaps here with 1 bit per replica?

Reply:
I'm not yet sure how we would determine that. Probably we could look at RPC activity, number of clients, etc to determine it. Ganesha has some ability to gather and present stats, but I haven't looked at them in detail.

To migrate a client off the server we'd need to push out a new fs_locations_info map to the clients, that did not contain the replica we'd want to remove from the cluster. They should then see that and move to an alternate address.

I've not looked in detail at the client code that handles this, but they should poll for a new fs_locations_info map periodically (based on an interval that the server advertises). So, migrating clients from one server to another would likely be something that takes on the order of minutes to an hour. Once they've all moved we could then decommission the old server.